django/contrib/admin/options.py

```diff
@@ -464,11 +465,11 @@ def lookup_allowed(self, lookup, value):
             # It is allowed to filter on values that would be found from local
             # model anyways. For example, if you filter on employee__department__id,
             # then the id value would be found already from employee__department_id.
-            if not prev_field or (
-                prev_field.is_relation
-                and field not in prev_field.path_infos[-1].target_fields
+            relation_parts.append(part)
+            if field is not model._meta.pk and not (
+                prev_field and part in getattr(prev_field, "to_fields", [])
             ):
-                relation_parts.append(part)
+                relation_parts_without_pk_and_to_fields.append(part)
```

Comment:
I'm not sure if this is correct, as it will allow for `restaurant__country` which should not be allowed :thinking: 
```diff
diff --git a/tests/modeladmin/tests.py b/tests/modeladmin/tests.py
index 3cf780c7e2..06c262f8aa 100644
--- a/tests/modeladmin/tests.py
+++ b/tests/modeladmin/tests.py
@@ -175,6 +175,9 @@ class ModelAdminTests(TestCase):
         self.assertIs(
             ma.lookup_allowed("restaurant__place__country", "test_value"), True
         )
+        self.assertIs(
+            ma.lookup_allowed("restaurant__country", "test_value"), False
+        )
 
     def test_field_arguments(self):
         # If fields is specified, fieldsets_add and fieldsets_change should
```

Reply:
Ah good spot, I will have a think :+1:

Reply:
So specifically the case you are talking about we have further up the file a comment
```python
            try:
                field = model._meta.get_field(part)
            except FieldDoesNotExist:
                # Lookups on nonexistent fields are ok, since they're ignored
                # later.
                break
```
So basically, it goes through the parts ["restaurant", "country"] sees that country does not exist and then breaks the loop and has decided this must be a local field filter

Then in `get_filters` it handles the `FieldDoesNotExist` error

What do you think? I could add a comment that lookupallowed has false positives when fields do not exist
`ma.lookup_allowed("nonexistant_field", "test_value")` returns True for example

Reply:
I tried to use:
```python
class WaiterAdmin(admin.ModelAdmin):
    list_filter = ["restaurant__place__country"]
```
in a sample project and it still crashes with `DisallowedModelAdminLookup` because a different lookup is passed to the `lookup_allowed()`, i.e. `restaurant__place__country__id__exact`:

![image](https://user-images.githubusercontent.com/2865885/227879666-6d3977d3-07ae-4b0e-9329-b669d5237a3d.png)

As far as I'm aware, a proper regression test should assert:
```python
self.assertIs(
    ma.lookup_allowed("restaurant__place__country__id__exact", "1"), True
)
```

Reply:
Well caught, I have pushed up a fix :+1: 

Just a note I need the extra array `relation_parts_without_to_fields`  for `admin_filters.tests.ListFiltersTests.test_fk_with_to_field` to pass, it would be nice not to have the extra array but I'm struggling to lose it completely :thinking:

Reply:
> it would be nice not to have the extra array but I'm struggling to lose it completely

Yeah, it's tricky, I need a blackboard for this :smiley: . One thing that I noticed is that we don't need `+ [part]` anymore as it only produces invalid lookups
```diff
diff --git a/django/contrib/admin/options.py b/django/contrib/admin/options.py
index 73c6e38933..554e5132d1 100644
--- a/django/contrib/admin/options.py
+++ b/django/contrib/admin/options.py
@@ -498,8 +498,6 @@ class BaseModelAdmin(metaclass=forms.MediaDefiningClass):
         return not {
             LOOKUP_SEP.join(relation_parts),
             LOOKUP_SEP.join(relation_parts_without_to_fields),
-            LOOKUP_SEP.join(relation_parts + [part]),
-            LOOKUP_SEP.join(relation_parts_without_to_fields + [part]),
         }.isdisjoint(valid_lookups)
 
     def to_field_allowed(self, request, to_field):
```

Reply:
I'm not sure if the logic is foolproof but I lost the extra array :tada: 

But after my recent change I needed the `+ [part]` again for this test `modeladmin.tests.ModelAdminTests.test_lookup_allowed_onetoone`